cachito/workers/tasks/general.py

```diff
@@ -49,6 +50,30 @@ def fetch_app_source(url, ref, request_id, gitsubmodule=False):
     bundle_dir = RequestBundleDir(request_id)
     log.debug("Extracting %s to %s", scm.sources_dir.archive_path, bundle_dir)
     shutil.unpack_archive(str(scm.sources_dir.archive_path), str(bundle_dir))
+    _enforce_sandbox(bundle_dir.source_root_dir)
+
+
+def _enforce_sandbox(repo_root):
+    """
+    Check that there are no symlinks that try to leave the cloned repository.
+
+    :param (str | Path) repo_root: absolute path to root of cloned repository
+    :raises ValidationError: if any symlink points outside of cloned repository
+    """
+    for dirpath, subdirs, files in os.walk(repo_root):
+        dirpath = Path(dirpath)
+
+        for entry in subdirs + files:
+            full_path = dirpath / entry
+            real_path = full_path.resolve()
+            try:
+                real_path.relative_to(repo_root)
+            except ValueError:
+                # This one cannot fail
+                relative_path = str(full_path.relative_to(repo_root))
+                raise ValidationError(
+                    f"The destination of {relative_path!r} is outside of cloned repository"
```

Comment:
Why not just use the error message from the `ValueError` exception?

Reply:
The error message would be `'<real path>' does not start with '<full path to bundle app dir>'`, e.g. `'/src' does not start with '/tmp/cachito-archives/bundles/temp/2/app'`

IMO, users will care more about the relative path to the problematic file in their repo, and I'd say they don't need to know what it resolved to.

Reply:
That's a great point. Your approach makes it easier to diagnose the issue.